libs/galley-types/src/Galley/Types/Teams.hs

```diff
@@ -149,13 +149,25 @@ rolePermissions :: Role -> Permissions
 rolePermissions role = Permissions p p where p = rolePerms role
 
 permissionsRole :: Permissions -> Maybe Role
-permissionsRole (Permissions p p') | p /= p' = Nothing
+permissionsRole (Permissions p p')
+  | p /= p' =
+    -- we never did use @p /= p'@ for anything, fingers crossed that it doesn't occur anywhere
+    -- in the wild.  but if it does, this implementation prevents privilege escalation.
+    let p'' = Set.intersection p p'
+     in permissionsRole (Permissions p'' p'')
 permissionsRole (Permissions p _) = permsRole p
```

Comment:
line 152 and line 158 have overlapping patterns.   Will haskell actually fall back when the guard doesn't match? or would we have to move line 158 into an `| otherwise = ` ?

Reply:
it does fall back, you've hit on something, this here is actually more clear and robust: f12e85c58fdcbc9b00507c148e8e8dffb8a78dd7